node/beacon-chain-handler/src/lib.rs

```diff
@@ -0,0 +1,74 @@
+extern crate beacon;
+extern crate chain;
+extern crate node_runtime;
+extern crate primitives;
+extern crate storage;
+
+use beacon::types::BeaconBlock;
+use chain::BlockChain;
+use node_runtime::{ApplyState, Runtime};
+use primitives::traits::{Block, Header};
+use primitives::traits::Signer;
+use primitives::types::ConsensusBlockBody;
+use primitives::types::SignedTransaction;
+use std::sync::Arc;
+use storage::StateDb;
+
+pub trait ConsensusHandler<B: Block, P>: Send + Sync {
+    fn produce_block(&self, body: ConsensusBlockBody<P>) -> B;
+}
+
+pub struct BeaconBlockProducer {
+    beacon_chain: Arc<BlockChain<BeaconBlock>>,
+    runtime: Runtime,
+    signer: Arc<Signer>,
+    state_db: Arc<StateDb>,
+}
+
+impl BeaconBlockProducer {
+    pub fn new(
+        beacon_chain: Arc<BlockChain<BeaconBlock>>,
+        runtime: Runtime,
+        signer: Arc<Signer>,
+        state_db: Arc<StateDb>,
+    ) -> Self {
+        BeaconBlockProducer {
+            beacon_chain,
+            runtime,
+            signer,
+            state_db,
+        }
+    }
+}
+
+pub type BeaconChainConsensusBlockBody = ConsensusBlockBody<Vec<SignedTransaction>>;
+
+impl ConsensusHandler<BeaconBlock, Vec<SignedTransaction>> for BeaconBlockProducer {
+    fn produce_block(&self, body: BeaconChainConsensusBlockBody) -> BeaconBlock {
```

Comment:
if we are trying to build real interface - we should remove signing from here, and it should just return back to txflow what they going to be signing on (hash of the block).
The constructed block with state should be put in some cache, until it's signed and only then added to the chain.

Reply:
If we implement future interface for the signer then this method can return future too and there is no need to store the unsigned block somewhere (technically, the unsigned block will temporarily live inside the stack of the produce_block function while its future is pending)